commown_shareholder_register/models/shareholder_register.py

```diff
@@ -0,0 +1,188 @@
+    ids = tuple({item[attr][0] for item in data})
+    return {obj.id: obj for obj in env[model].search([("id", "in", ids)])}
+
+
+def _concatenate_address(partner):
+    address_chunks = [
+        partner.street2,
+        partner.street,
+        partner.zip,
+        partner.country_id.name,
+    ]
+    result = ""
+    for chunk in address_chunks:
+        if chunk:
+            result += " " + chunk
+    return result
+
+
+class ShareholderRegister(models.TransientModel):
+    _name = "commown_shareholder_register.register"
+    _description = "Utility class to compute a shareholder register"
+
+    date = fields.Datetime(
+        string="Date",
+        help="Date of the register",
+    )
+    report = fields.Binary("Shareholder Register", readonly=True)
+    report_name = fields.Char(string="Filename", size=256, readonly=True)
+
+    def get_shareholders(self):
+        data = self.env["account.move.line"].read_group(
+            [
+                ("account_id.code", "=like", "1013%"),
+                ("partner_id", "!=", False),
+                ("date", "<", self.date),
+                ("partner_id.active", "=", True),
+            ],
+            ["account_id", "partner_id", "balance:sum"],
+            ["account_id", "partner_id"],
+            lazy=False,
+        )
+
+        partners = _by_id(self.env, data, "partner_id", "res.partner")
+        accounts = _by_id(self.env, data, "account_id", "account.account")
+
+        result = {
+            "warnings": [],
+            "partners": {},
+            "colleges": {},
+            "total": {
+                "balance": 0,
+            },
+        }
+
+        for item in data:
+            account = accounts[item["account_id"][0]]
+            partner = partners[item["partner_id"][0]]
+
+            if -item["balance"] >= SHARE_TYPE_DATA[account.code]["min_share_amount"]:
+                result["partners"].setdefault(
+                    partner,
+                    {
+                        "total": 0,
+                        "college": {"college_rank": 0},
+                        "address": _concatenate_address(partner),
+                        "phone": partner.phone or "Not Available",
+                    },
+                )["total"] -= item["balance"]
+                if (
+                    result["partners"][partner]["college"]["college_rank"]
+                    < SHARE_TYPE_DATA[account.code]["college_rank"]
+                ):
+                    result["partners"][partner]["college"] = SHARE_TYPE_DATA[
+                        account.code
+                    ]
+
+                result["total"]["balance"] -= item["balance"]
+
+            elif item["balance"] > 0:
+                result["warnings"].append(
+                    _("The partner %s has a negative share number") % partner.name
+                )
+
+        for partner in result["partners"].keys():
+            result["partners"][partner]["share_number"] = (
```

Comment:
for partner, partner_data in result["partners"].items() permettrait de réutiliser partner_data.